c2corg_api/views/waypoint.py

```diff
@@ -34,9 +33,96 @@ def get(self):
     def collection_post(self):
         waypoint = schema_waypoint.objectify(self.request.validated)
 
+        # TODO additional validation: at least one culture, only one instance
+        # for each culture
+
         DBSession.add(waypoint)
         DBSession.flush()
 
         self._create_new_version(waypoint)
 
         return to_json_dict(waypoint, schema_waypoint)
+
+    @view(schema=schema_update_waypoint, validators=validate_id)
+    def put(self):
+        id = self.request.validated['id']
+        waypoint_in = \
+            schema_waypoint.objectify(self.request.validated['document'])
+        self._check_document_id(id, waypoint_in.document_id)
+
+        waypoint = self._get_waypoint(id)
+        self._check_versions(waypoint, waypoint_in)
+        old_versions = waypoint.get_versions()
+        waypoint.update(waypoint_in)
+
+        DBSession.merge(waypoint)
+        DBSession.flush()
+
+        (update_type, changed_langs) = \
+            self._check_update_type(waypoint, old_versions)
+        self._update_version(
+            waypoint, self.request.validated['message'], update_type,
+            changed_langs)
+
+        return to_json_dict(waypoint, schema_waypoint)
+
+    def _get_waypoint(self, id, culture=None):
+        """Get a waypoint with either a single locale (if `culture is given)
+        or with all locales.
+        If no waypoint exists for the given id, a `HTTPNotFound` exception is
+        raised.
+        """
+        if not culture:
+            waypoint = DBSession. \
+                query(Waypoint). \
+                filter(Waypoint.document_id == id). \
+                options(joinedload(Waypoint.locales)). \
+                first()
+        else:
+            waypoint = DBSession. \
+                query(Waypoint). \
+                join(Waypoint.locales). \
+                filter(Waypoint.document_id == id). \
+                options(contains_eager(Waypoint.locales)). \
+                filter(DocumentLocale.culture == culture). \
+                first()
+
+        if not waypoint:
+            raise HTTPNotFound('document not found')
```

Comment:
`_get_waypoint` is also used in `WaypointRest.get` and I didn't want to repeat the `...raise ...` part.